### DISCUSSION TITLE
Unit Test Parsing

### DISCUSSION THREAD

**Discussion by ballcoach12** [eclipse-langium/langium#340] [Q&A] [answered]
Is there an example that exists to demonstrate how to unit test parsing of concrete syntax? I'm accustomed to using injecting a ParseHelper<DomainModelRoot> with Xtext to invoke lexing and parsing, and then validating the result, but I assume that I'd need to write my own ParseHelper for Langium, but I'm not really sure where to start.

Greg

**Comment by msujew** [comment, marked as answer]
Hi @ballcoach12,

we actually provide a `parseHelper` function, but it's exported from the `langium/lib/test` path of the framework.

You can see how it works in action:

https://github.com/langium/langium/blob/850f92226032aaa74a4695d5d8847a9a3026332f/packages/langium/test/parser/token-builder.test.ts#L27

It basically creates a curried function based on an instance of `LangiumServices`. Afterwards you can throw in any text and get a `LangiumDocument` back.

**Reply by ballcoach12** [reply]
Thank you, Mark. I am getting some errors related to missing types (i.e. LangiumSharedServices) when I try to use the function. 

I  pulled the master branch of langium and have tried to build the statemachine example since its module file has a function that looks similar to what I would need for my test. However, I am unable to build the example either, due to the same missing types errors.  I get errors like this:

> Module '"langium"' has no exported member 'DefaultSharedModuleContext'.ts(2305)

I have run npm install and npm run langium:generate on the statemachine project. Is there something that I'm doing wrong?

**Reply by ballcoach12** [reply]
Also, when I run 'npm run langium:generate' on the statemachine project, an error is returned stating: 

Error: Your Langium configuration is invalid.
--> instance[0] is not allowed to have the additional property "projectName"
--> instance[0] is not allowed to have the additional property "languages"  
--> instance[0] requires property "grammar"
--> instance[0] requires property "fileExtensions"
--> instance[0] requires property "languageId"

**Reply by ballcoach12** [reply]
OK. So it looks like 0.2.0 doesn't have the shared services stuff in it, which would make sense why my code doesn't  compile against that version. My guess is that I need to point my language's dependencies at the langium source code that I pulled onto my machine.

**Reply by msujew** [reply]
I see. Yeah mix & match of different langium libraries/versions is definitely not recommended, since everything is highly unstable right now and most APIs are changing with every release.

I assume you were somehow able to get it to work, since you selected my answer?

**Reply by ballcoach12** [reply]
Yes, I did get it to work. I went back to Langium v 0.2.0 and adjusted my tests accordingly. 

I have to admit that I don't quite comprehend how the DI works, though. I'm accustomed to frameworks like Inversify, and the service approach is new to me. Perhaps it's just differences in terminology, though.  If there is a document that explains the DI system in Langium, it would probably help me better grasp how things work.
